pkg/jsonmerge/object.go

```diff
@@ -0,0 +1,73 @@
+package jsonmerge
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+
+	"dario.cat/mergo"
+)
+
+type objectMerger struct {
+	io.Writer
+	dst map[string]interface{}
+}
+
+// NewObjectMerger creates a Merger for JSON objects.
+func NewObjectMerger(w io.Writer) Merger {
+	return &objectMerger{
+		Writer: w,
+		dst:    make(map[string]interface{}),
+	}
+}
+
+func (merger *objectMerger) NewPage(r io.Reader, isLastPage bool) io.ReadCloser {
+	return &objectMergerPage{
+		merger: merger,
+		Reader: r,
+	}
+}
+
+func (merger *objectMerger) Close() error {
+	// Marshal to JSON and write to output.
+	buf, err := json.Marshal(merger.dst)
+	if err != nil {
+		return err
+	}
+
+	_, err = merger.Writer.Write(buf)
+	return err
+}
+
+type objectMergerPage struct {
+	merger *objectMerger
+
+	io.Reader
+	buffer bytes.Buffer
+}
+
+func (page *objectMergerPage) Read(p []byte) (int, error) {
+	// Read into a temporary buffer to be merged and written later.
+	p = make([]byte, len(p))
+	n, err := page.Reader.Read(p)
+	if err != nil {
+		return 0, err
+	}
+	if n == 0 {
+		return 0, io.EOF
+	}
+
+	_, err = page.buffer.Write(p[:n])
+	return 0, err
+}
+
+func (page *objectMergerPage) Close() error {
+	var src map[string]interface{}
+
+	err := json.Unmarshal(page.buffer.Bytes(), &src)
+	if err != nil {
+		return err
```

Comment:
Given that this can happen either because:
 * You didn't provide JSON
 * You didn't fully read the input buffer

Do you think it's worth first doing a check for whether there are still bytes to read in the input buffer and providing a useful error around that? Because we're using `io.Copy` in the api I'm not sure this can ever happen but maybe?

Reply:
This wasn't intended to be go routine-safe. I'm not aware of much IO in the std library that is. By the time `Close` is called there should be no current or further calls to `Read` expected.

Reply:
I'm not talking about goroutines here. Here's a test that demonstrates what I mean:

```go
func TestObjectMerger_notFullyRead(t *testing.T) {
	w := &bytes.Buffer{}
	merger := NewObjectMerger(w)

	r1 := bytes.NewBufferString(`{}`)
	p1 := merger.NewPage(r1, true)
	n, err := p1.Read(make([]byte, 1)) // only reading 1 byte

	require.NoError(t, err)
	assert.Equal(t, 0, n)

	require.NoError(t, p1.Close())
}
```

```
--- FAIL: TestObjectMerger_notFullyRead (0.00s)
    /Users/williammartin/workspace/go-gh/pkg/jsonmerge/object_test.go:87: 
        	Error Trace:	object_test.go:87
        	Error:      	Received unexpected error:
        	            	unexpected end of JSON input
        	Test:       	TestObjectMerger_notFullyRead
```

This fails because we only `Read` 1 byte of the page and couldn't unmarshal a valid JSON document.

This leans into my concern about shipping this as its own package as-is because it's tightly coupled to the expectations of the calling code (i.e. that you consume every byte which we _know_ happens because we `io.Copy` but that's not required by someone calling `Read`) and that isn't documented anywhere on the package interface.

I'm suggesting in my review comment that if there are still bytes in the `Reader` when we call `Close`, we could help users of this package by pointing them to this fact, rather than having them scratching their head about the structure of the string they passed in. It seems like a nice quality of life improvement.

Yes, it's obvious that you need to ensure that the entire document is consumed to have a valid JSON structure but especially if you happen to be some degrees of nested different `Reader`s away it's not always obvious what's happening. And I know you're a Rust developer and value the compiler being so helpful 😬 

---

Also, though I care less about this one, it's possible that an _invalid_ JSON string can be consumed and merged unexpectedly by not emptying the `Reader`:

```
func TestObjectMerger_notFullyReadInvalidObject(t *testing.T) {
	w := &bytes.Buffer{}
	merger := NewObjectMerger(w)

	r1 := bytes.NewBufferString(`{}}`)
	p1 := merger.NewPage(r1, true)
	n, err := p1.Read(make([]byte, 2))

	require.NoError(t, err)
	assert.Equal(t, 0, n)

	require.Error(t, p1.Close())
}
```

Though I haven't looked as closely, I would assume the same is true for the `arrayMerger`.

Reply:
I see what you mean. But how would we ensure there's more to read? Just try reading from the underlying `io.Reader` again and make sure we get 0 bytes and/or `io.EOF`?

That said, I appreciate the concern about putting this in cli/go-gh and making it more "general purpose". I'm fine with porting it to cli/cli and keeping it internal. If ever there's a need - like many of the things including the `tableprinter` I helped improve and migrate out - to move it out, we can deal with generalizing it then.

Sound good? I could still make some improvements to better guard against possible issues, but making this internal and controlling the use patterns would mitigate several of these concerns.

Reply:
If we moved this to `cli/internal` then I would be comfortable not including this change since we know we're in control of the usage.

> Just try reading from the underlying io.Reader again and make sure we get 0 bytes and/or io.EOF?

That's what I would have tried to do, perhaps naively.